src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1100,16 +1141,20 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
         /// <param name="other"></param>
         /// <param name="whenTrueValues">The possible values of test.Input when <paramref name="test"/> has succeeded.</param>
         /// <param name="whenFalseValues">The possible values of test.Input when <paramref name="test"/> has failed.</param>
+        /// <param name="precondition">The pre-condition that the implied status of the <paramref name="other"/> test is dependent upon.</param>
```

Comment:
nit: let's add xml doc for `sideeffect` too

Reply:
I'm thinking that sideeffect should be set even if there's no preconditions (in case inputs are not identical in any ways). If I couldn't come up with a test case for that we can simplify things in a few places you mentioned below.

Reply:
> I'm thinking that sideeffect should be set even if there's no preconditions (in case inputs are not identical in any ways).

I think the `sideeffect` is only useful when the tests are related via `precondition`.
There's two scenarios:
- tests are related (via `precondition`) so we may be able to skip `other`. `sideeffect` is part of that (we need it to figure out remaining values since we're not running `other`)
- tests are unrelated, so we'll have to evaluate `other` and that will give us the remaining values by itself (shouldn't merge with any other set, so no sideeffect/assignment/merge).

Does that make sense?

Reply:
Ok here's my reasoning to always emit the assignment node: 

When inputs are not identical, from this point on we have determined that they are practically the same so we can avoid fetching the second evaluation for better codegen (https://github.com/dotnet/roslyn/pull/53891#discussion_r688394531)